src/optimizer/filter_combiner.cpp

```diff
@@ -987,4 +994,201 @@ ValueComparisonResult CompareValueInformation(ExpressionValueInformation &left,
 	}
 }
 
+void FilterCombiner::LookUpConjunctions(Expression *expr) {
+	if (expr->GetExpressionType() == ExpressionType::CONJUNCTION_OR) {
+		auto root_or = (BoundConjunctionExpression *)expr;
+		for (auto &or_to_push : ors_to_pushdown) {
+			if (or_to_push->root_or->Equals(root_or)) {
+				return;
+			}
+		}
+
+		auto or_to_push = make_unique<OrToPush>();
+		or_to_push->root_or = root_or;
+		or_to_push->cur_conjunction = root_or;
+		ors_to_pushdown.emplace_back(move(or_to_push));
+
+		BFSLookUpConjunctions(root_or);
+		return;
+	}
+
+	// Verify if the expression has a column already pushed down by other OR expression
+	VerifyOrsToPush(*expr);
+}
+
+void FilterCombiner::BFSLookUpConjunctions(BoundConjunctionExpression *conjunction) {
+	if (VerifyEarlyStopPushdown()) {
+		return;
+	}
+
+	vector<BoundConjunctionExpression *> conjunctions_to_visit;
+
+	for (auto &child : conjunction->children) {
+		switch (child->GetExpressionClass()) {
+		case ExpressionClass::BOUND_CONJUNCTION: {
+			auto child_conjunction = (BoundConjunctionExpression *)child.get();
+			conjunctions_to_visit.emplace_back(child_conjunction);
+			break;
+		}
+		case ExpressionClass::BOUND_COMPARISON: {
+			UpdateConjunctionFilter((BoundComparisonExpression *)child.get(), *child);
+			break;
+		}
+		default: {
+			SetEarlyStopPushdown(true);
```

Comment:
This stateful approach of using a function (`SetEarlyStopPushdown`) to set a flag somewhere that aborts the pushdown seems quite error-prone. Could we not just return from this function whether or not the pushdown has succeeded, and not push down the OR node at all if the pushdown has not succeeded (for whatever reason)?

Reply:
I think, at this point, it is possible to return and not push down the OR node.